### PR TITLE
Ability to add/edit campaigns

### PR DESCRIPTION
**Pull request by tomverran**

### CODE REVIEW

File: app/controllers/CampaignController.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+class CampaignController(service: JsonDynamoService[Campaign, Future]) {
+
+  def all = Action.async {
+    service.all.map(campaigns => Ok(Json.toJson(campaigns)))
+  }
+
+  def upsert = Action.async { request =>
+    request.body.asJson.map { json =>
+      println(json)
```
Comment (paulbrown1982) [inline app/controllers/CampaignController.scala]: Remember to remove this.

File: frontend/src/controllers/EditCampaignController.es6 [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    fetchCampaign(code) {
+        let campaign = this.service.all();
+        campaign.then(r => r.filter(v => v.code == code)[0]).then(r => {
+            this.$scope.campaign = r ? r : {name: "", code: ""}
+        });
+    }
+
+    save(campaign) {
+        this.service.save(campaign);
+        console.log(campaign, "is saved now");
```
Comment (paulbrown1982) [inline frontend/src/controllers/EditCampaignController.es6]: Reminder to remove this logging, or to raise a Trello ticket to cover creating a proper logging strategy.

File: frontend/src/controllers/StageController.es6 [hunk 3]
```diff
@@ -3,16 +3,17 @@ const STAGE = 'stage';
...
     /* @ngInject */
-    constructor($state, $scope, $cookies) {
-        this.$scope = $scope;
+    constructor($state, $rootScope, $scope, $cookies) {
         this.$state = $state;
+        $rootScope.stage = {name: "PROD"};
+        this.$scope = $scope;
         this.$cookies = $cookies;
-        this.$scope.stage = this.$cookies.get(STAGE);
+        this.$scope.stage.name = this.$cookies.get(STAGE);
```
Comment (paulbrown1982) [inline frontend/src/controllers/StageController.es6]: Request we rename 'stage' to 'environment' to match the UI.

**Comment by paulbrown1982** [comment]
👍 It's looking really great!

I've captured the above comments on the feature/bug list, plus the following:

- Can the side bar auto-close after clicking an environment.
- Can we print the campaign in a breadcrumb when viewing the promotion edit page

**Comment by tomverran** [comment]
I agree with these comments, though we refer to the values DEV / UAT / PROD as stage in membership / subscriptions frontend. I guess the relationship between the terms is that the stage determines the environment

Anyway I'm going to merge this so we can continue working off this code

**Pull request merged**